lib/puppet/util/selinux.rb

```diff
@@ -49,7 +49,10 @@ def get_selinux_default_context(file)
     end
 
     retval = Selinux.matchpathcon(file, mode)
-    Selinux.matchpathcon_fini
+    # The implementation of matchpathcon/matchpathcon_fini in RedHat before version 7 was very slow
```

Comment:
@MikaelSmith I've been looking in to this. It's actually matchpathcon_init (called automatically by matchpathcon) that's so slow.  EL7 introduced 'file_contexts.bin files with precompiled regular expressions'.
For a moment, I thought using matchpathcon_init_prefix was a viable solution but it is still nowhere near fast enough depending on the path of the file.  It made your `puppet apply -e 'notify {"hello":}` test much faster again, but was still really slow in other cases.

Even though matchpathcon_init_prefix will accept a full path, it only uses the first component (eg '/opt') and a large chunk of the context file still ends up being loaded, and this is slow.

Reply:
matchpathcon_init is presumably what fixed the issue. Could we instead do matchpathcon_init and matchpathcon_free in external from the file provider? If we need to call it once per run, I'd be comfortable doing that somewhere else.

It must persist across runs because the Puppet service applies settings, which will initialize the cache.

Reply:
Somewhere in https://github.com/puppetlabs/puppet/blob/master/lib/puppet/agent.rb#L42-L55 could be an appropriate place to matchpathcon_fini.

Reply:
Once per run would be an improvement on the current behaviour, but if a package containing new selinux policy is installed, (or maybe an exec run), it's still going to take 2 puppet runs for the correct contexts to be set.

Has this PR (fixing it for EL7 only) been ruled out completely, or just for the current releases?
Perhaps if this only went into master and wasn't backported to 4.10.x that'd be acceptable?

Maybe there's a better solution we've not thought of yet.  Don't call matchpathcon_fini if the last resource applied was also a File resource?  Not sure that's even possible though and I'm scared of corner cases.

Reply:
@branan did you have any thoughts on this?

p.s. might take him a day or two to get around to this.

Reply:
For what it's worth, I did also manage to get the replacement functions selabel_open, selabel_lookup and selabel_close working.  Just like matchpathcon_init, selabel_open was fast on EL7 but awful on EL6.  If you do end up going for some EL7 specific solution, perhaps it'd be an opportunity to switch to the new functions.